### ISSUE TITLE
question: word callbacks?

### ISSUE THREAD

**Issue by pazos** [koreader/koreader#12252]
Saw a question in https://www.mobileread.com/forums/showpost.php?p=4439943&postcount=1

> add a note to a particular word and then have that note show up for each mention of the word in the book

I would love a feature like that, so I'm thinking about implementing it myself. I have no clue yet.

I was thinking about a `word callback` where an action is pegged to every appearance of a word. The action is meant to display enhanced info.

pinging @poire-z, @hius07 as you know how crengine integration works, highlights et al. **I don't**. So maybe this is a bluff and the amount and kind of work is too much for me :/

**Comment by hius07** [comment]
It looks similar to Kindle WordWise
https://github.com/koreader/koreader/issues/3044#issuecomment-748504419

**Comment by poire-z** [comment]
I think it's not worth (and too much non-crengine-business) to integrate it into crengine itself.
And if not doing it via crengine, you get to have it work with PDF too.
Thinking 25 seconds, I see 2 ways to go at that:
- on each page turn, get the full text of the page, and search your wordS in that. I don't really know if we can get the x/y/w/h of each rectangle at this point, so it may not be a good idea
- at start of document, your plugin (because it can only be a plugin :) I'd rather not have that monstruosity in core :)) would search (like our new search all occurences) all your wordS (but then again, part of words, or anchores with word boundaries), and get a list of "highlights". Either made part of our regular highlights (so, noise in them), or another highlight-like list, that would be rendered apart (but just like our highlights, by readerview itself).

Or may be combine these 2 ideas: quick search in the plain text, and if matches, "search all" or something limited to looking into this page.
Or may be directly just "search all" in the current page (if we have that, dunno).
Or may be I don't know :) there's nothing really obvious- and the above ideas quite intrude a lot into the highlight and search core features :/ so, monstruosity it seems.

If a "word callback", depends on the amount of info you want with it. If just a callback to say "there's some in the page", getting the page full text could just be enough, and doable in a plugin.

**Comment by pazos** [comment]
thanks for your time!

> It looks similar to Kindle WordWise

Yup, pretty much.

Like the screenshot in https://github.com/koreader/koreader/issues/3044#issuecomment-646820645

But "enhanced words" are displayed as highlighted words (with a different pattern) and will popup the enhanced info when pressed, using common widgets.

Then, instead of a quick definition of a word, you get plenty of space to describe a character, a place, an event or some other recurring word over a document.

My idea is to use the current highlights action menu as the entry point to enhance a word

**Comment by pazos** [comment]
Thanks for your time, @poire-z!

>I think it's not worth (and too much non-crengine-business) to integrate it into crengine itself.
And if not doing it via crengine, you get to have it work with PDF too.

Agree with the first statement. Disagree with the second. I'm fine if word enhancer works only on crdocuments :) 

> at start of document, your plugin (because it can only be a plugin :) I'd rather not have that monstruosity in core :)) would search (like our new search all occurences) all your wordS (but then again, part of words, or anchores with word boundaries), and get a list of "highlights". Either made part of our regular highlights (so, noise in them), or another highlight-like list, that would be rendered apart (but just like our highlights, by readerview itself).

Yay. A plugin. It shouldn't be in core but it shouldn't be a monstruosity either.
Keep enhanced words as a highlight like list, rendered by readerviewer, sounds sensible to me.

Enhanced words belong to a person, that uses its own lore to write descriptions, so they shouldn't be tied to documents. They also shouldn't be applied to new documents. A new document won't execute any related codepath until the first word in a document is enhanced. 

So I think I need to store word - enhancement pairs in a single file. Each enhanced document should have an array of words to be enhanced. I think that's the only bit of info that goes into current metadata.

> Or may be I don't know :) there's nothing really obvious- and the above ideas quite intrude a lot into the highlight and search core features :/ so, monstruosity it seems.

I don't get the monstruosity part :)

**Comment by Frenzie** [comment]
> Disagree with the second. I'm fine if word enhancer works only on crdocuments :)

It's just saying you'll get it "for free".

**Comment by poire-z** [comment]
> I don't get the monstruosity part :)

You will :)

**Comment by Commodore64user** [comment]
Another idea would be a sort x-ray but inverted. On kindle x-ray is essentially like a dictionary but for characters and places mentioned in a particular book (infamous for providing spoilers though), you could do the same but where you provide the info to whatever word you like.

so every time you select and long press the word or words you have saved, you get the dictionary pop up with the info you saved.

**Comment by jonnyl2** [comment]
I have been using Fulltext search for this and similar use cases. Incidentally, I just wrote up some related FR/suggestions I am shortly going to post as a separate Issue.

**Comment by Commodore64user** [comment]
> I have been using Fulltext search for this and similar use cases. Incidentally, I just wrote up some related FR/suggestions I am shortly going to post as a separate Issue.

Full text search and #11454 would be “chef’s kiss”.

**Comment by hius07** [comment]
An idea of a "word comment".

(1) Do a fulltext search of a "word to be commented".
In the All results list press the upper left menu icon.

<img width="400" height="540" alt="Image" src="https://github.com/user-attachments/assets/a0bf58ad-6fa3-4642-b93e-d85f0317e5a1" />

-----

(2) Add a comment.

<img width="400" height="540" alt="Image" src="https://github.com/user-attachments/assets/f0315791-f3be-48d6-b0e1-b0a095e1cb83" />

-----

(3) Every "word" in the book is now highlighted (new "dashed-line" style is proposed)

<img width="400" height="540" alt="Image" src="https://github.com/user-attachments/assets/2252dc2f-2100-449a-b223-013f344b5778" />

-----

(4) Pressing on any "word" shows the comment.

<img width="400" height="540" alt="Image" src="https://github.com/user-attachments/assets/86dc9585-90b4-4d03-aab0-8bc2207d53d3" />

-----

Reasons to start from the Search:
-(even hidden) search always must be done to highlight all entries in a book
-and more important, in languages with declension of nouns a user could search for a word root only, and get all entries of this word commented.
I think it's more convenient than to highlight the whole word, then edit the highlight to unhighlight the suffix, and then comment (and do the fulltext search in any case).

**Comment by hius07** [comment]
@Frenzie @poire-z what do you think about this feature?

The use case from mobileread:
I'm reading War and Peace and having trouble keeping track of all the characters, so it would be great if I could write up a brief description of each of them and then attach that description as a note to each mention of their name in the text.

I see it as based on the annotations infrastructure, with huge rework of the Bookmark module.

**Comment by poire-z** [comment]
Not a customer for the feature, and not a fan of answering every user request :)
The fast that it is invoked from Search makes this little feature not that intrusive.
But if you are going to create **real** highlight items for **each** occurence of these words, you may end up with tons of items and huge metadata.epub.lua, and lots of code to discriminate normal highlights from these fake highlights, which won't be fun to read and maintain (+ exporters & KOHighlights will need to ignore it).

**Comment by hius07** [comment]
Exactly, "comment" annotations are created to draw them and to respond to a tap.

**Comment by jonnyl2** [comment]
Are the search-term comments stored by book? If so, one drawback I could see would be for readers who read lots of series, where the same characters appear in multiple books. The search terms and comments would have to be entered for each book separately.

I was thinking that a possible approach would be to let users create a (global) custom dictionary with headwords and definitions. Headwords in this dictionary would be automatically underlined in the text and the definitions viewable as comments (as shown above). It could also be useful for language learners or students of subjects with many arcane terms.

I don't know how or if this could be integrated with search and annotations, though.

**Comment by poire-z** [comment]
Then I don't like it - not worth the kludge for such a peripheral request, annotations/bookmarks/highlights are already complicated enough.
People can use notebooks and write their own thoughts and train their memory.

**Comment by Frenzie** [comment]
Intuitively I'd say to simply search through the text on the current page to apply certain "wordwise"/"word annotation"/whatever things. Involving annotations sounds strange to me, except in the sense that they could be part of the annotations object on the storage side.

Also note that https://github.com/smartscripts-nl/dynamic-xray exists. I don't know what it does (because I prefer to use paper for notes) but it may provide some ideas.
